File: android/src/main/java/com/ymrabtipacks/flutter_azimuth/FlutterAzimuthPlugin.java
package com.ymrabtipacks.flutter_azimuth;

import androidx.annotation.NonNull;
import android.annotation.SuppressLint;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import io.flutter.plugin.common.EventChannel;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import io.flutter.plugin.common.PluginRegistry.Registrar;

import static android.content.Context.SENSOR_SERVICE;
/** FlutterAzimuthPlugin */
@SuppressLint("NewApi")
public class FlutterAzimuthPlugin implements FlutterPlugin, MethodCallHandler, SensorEventListener, EventChannel.StreamHandler {

  private FlutterPluginBinding flutterpluging;
  int mAzimuth;
  int previousAz=0;
  private int sensorType;
  private boolean haveSensor = false, haveSensor2 = false;
  private SensorManager mSensorManager;
  private Sensor mRotationV, mAccelerometer, mMagnetometer;
  float[] rMat = new float[9];
  float[] orientation = new float[3];
  private float[] mLastAccelerometer = new float[3];
  private float[] mLastMagnetometer = new float[3];
  private boolean mLastAccelerometerSet = false;
  private boolean mLastMagnetometerSet = false;
  private EventChannel.EventSink mEventSink;

  private MethodChannel channel;


  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    flutterpluging = flutterPluginBinding;
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "checkDeviceSensors");
    channel.setMethodCallHandler(this);


    final EventChannel eventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "azimuthStream");
    eventChannel.setStreamHandler(this);

  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getSensors") && android.os.Build.VERSION.SDK_INT >=  android.os.Build.VERSION_CODES.KITKAT_WATCH) {
        sensorType = checkSensors();
        result.success(sensorType);
    } else {
      result.notImplemented();
    }
  }

  @SuppressLint("NewApi")
  private int checkSensors() {
    mSensorManager = (SensorManager) (flutterpluging.getApplicationContext().getSystemService(SENSOR_SERVICE));
    if (mSensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) == null) {
      if ((mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) == null) || (mSensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) == null)) {
        return 0;
      } else {
        return 1;
      }
    } else {
      return 2;
    }

  }


  @SuppressLint("NewApi")
  @Override
  public void onSensorChanged(SensorEvent event) {
    if (event.sensor.getType() == Sensor.TYPE_ROTATION_VECTOR) {
      SensorManager.getRotationMatrixFromVector(rMat, event.values);
      mAzimuth = (int) (Math.toDegrees(SensorManager.getOrientation(rMat, orientation)[0]) + 360) % 360;
    }

    if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
      System.arraycopy(event.values, 0, mLastAccelerometer, 0, event.values.length);
      mLastAccelerometerSet = true;
    } else if (event.sensor.getType() == Sensor.TYPE_MAGNETIC_FIELD) {
      System.arraycopy(event.values, 0, mLastMagnetometer, 0, event.values.length);
      mLastMagnetometerSet = true;
    }
    if (mLastAccelerometerSet && mLastMagnetometerSet) {
      SensorManager.getRotationMatrix(rMat, null, mLastAccelerometer, mLastMagnetometer);
      SensorManager.getOrientation(rMat, orientation);
      mAzimuth = (int) (Math.toDegrees(SensorManager.getOrientation(rMat, orientation)[0]) + 360) % 360;
    }

    mAzimuth = Math.round(mAzimuth);

    if(mEventSink!=null)
      mEventSink.success(mAzimuth);

    previousAz = mAzimuth;

  }

  @Override
  public void onAccuracyChanged(Sensor sensor, int accuracy) {

  }

  @SuppressLint("NewApi")
  public void start() {

    if (sensorType==1) {

        mAccelerometer = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        mMagnetometer = mSensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD);
        haveSensor = mSensorManager.registerListener(this, mAccelerometer, SensorManager.SENSOR_DELAY_UI);
        haveSensor2 = mSensorManager.registerListener(this, mMagnetometer, SensorManager.SENSOR_DELAY_UI);

    }
    else
      if(sensorType==2){
      mRotationV = mSensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
      haveSensor = mSensorManager.registerListener(this, mRotationV, SensorManager.SENSOR_DELAY_UI);
    }
  }

  @SuppressLint("NewApi")
  public void stop() {
    if (haveSensor) {
      mSensorManager.unregisterListener(this, mRotationV);
    }
    else {
      mSensorManager.unregisterListener(this, mAccelerometer);
      mSensorManager.unregisterListener(this, mMagnetometer);
    }
  }

  @Override
  public void onListen(Object o, EventChannel.EventSink eventSink) {
    mEventSink = eventSink;
    start();
  }

  @Override
  public void onCancel(Object o) {
    stop();
  }
}

File: android/src/main/kotlin/com/ymrabtipacks/flutter_azimuth/FlutterAzimuthPlugin.kt
package com.ymrabtipacks.flutter_azimuth

import android.annotation.SuppressLint
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterAzimuthPlugin */
@SuppressLint("NewApi")
class FlutterAzimuthPlugin : FlutterPlugin, MethodCallHandler, SensorEventListener, EventChannel.StreamHandler {

  private lateinit var pluginBinding: FlutterPlugin.FlutterPluginBinding
  private var sensorManager: SensorManager? = null
  private var sensorType: Int = 0
  private var rotationSensor: Sensor? = null
  private var accelerometer: Sensor? = null
  private var magnetometer: Sensor? = null
  private var hasPrimarySensor: Boolean = false
  private var hasSecondarySensor: Boolean = false
  private var eventSink: EventChannel.EventSink? = null
  private var methodChannel: MethodChannel? = null

  private val rotationMatrix = FloatArray(9)
  private val orientation = FloatArray(3)
  private val lastAccelerometer = FloatArray(3)
  private val lastMagnetometer = FloatArray(3)
  private var lastAccelerometerSet = false
  private var lastMagnetometerSet = false
  private var azimuth: Int = 0

  override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    pluginBinding = binding
    methodChannel = MethodChannel(binding.binaryMessenger, "checkDeviceSensors").also {
      it.setMethodCallHandler(this)
    }

    EventChannel(binding.binaryMessenger, "azimuthStream").setStreamHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel?.setMethodCallHandler(null)
    methodChannel = null
    stop()
    eventSink = null
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "getSensors" && Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
      result.success(checkSensors())
    } else {
      result.notImplemented()
    }
  }

  @SuppressLint("ServiceCast")
  private fun ensureSensorManager(): SensorManager {
    if (sensorManager == null) {
      sensorManager = pluginBinding.applicationContext.getSystemService(SensorManager::class.java)
    }
    return sensorManager!!
  }

  private fun checkSensors(): Int {
    val manager = ensureSensorManager()
    sensorType = when {
      manager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) != null -> 2
      manager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null && manager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) != null -> 1
      else -> 0
    }
    return sensorType
  }

  override fun onSensorChanged(event: SensorEvent) {
    when (event.sensor.type) {
      Sensor.TYPE_ROTATION_VECTOR -> {
        SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values)
        azimuth = Math.toDegrees(SensorManager.getOrientation(rotationMatrix, orientation)[0].toDouble()).toInt()
      }

      Sensor.TYPE_ACCELEROMETER -> {
        System.arraycopy(event.values, 0, lastAccelerometer, 0, event.values.size)
        lastAccelerometerSet = true
      }

      Sensor.TYPE_MAGNETIC_FIELD -> {
        System.arraycopy(event.values, 0, lastMagnetometer, 0, event.values.size)
        lastMagnetometerSet = true
      }
    }

    if (lastAccelerometerSet && lastMagnetometerSet) {
      SensorManager.getRotationMatrix(rotationMatrix, null, lastAccelerometer, lastMagnetometer)
      SensorManager.getOrientation(rotationMatrix, orientation)
      azimuth = Math.toDegrees(orientation[0].toDouble()).toInt()
    }

    azimuth = ((azimuth % 360) + 360) % 360

    eventSink?.success(azimuth)
  }

  override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
    // No-op
  }

  private fun resetSensorState() {
    lastAccelerometerSet = false
    lastMagnetometerSet = false
    hasPrimarySensor = false
    hasSecondarySensor = false
  }

  private fun start() {
    val manager = ensureSensorManager()
    if (sensorType == 0) {
      checkSensors()
    }

    resetSensorState()

    when (sensorType) {
      1 -> {
        accelerometer = manager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        magnetometer = manager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
        hasPrimarySensor = manager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_UI)
        hasSecondarySensor = manager.registerListener(this, magnetometer, SensorManager.SENSOR_DELAY_UI)
      }

      2 -> {
        rotationSensor = manager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
        hasPrimarySensor = manager.registerListener(this, rotationSensor, SensorManager.SENSOR_DELAY_UI)
      }
    }
  }

  private fun stop() {
    val manager = sensorManager ?: return
    if (hasPrimarySensor || hasSecondarySensor) {
      rotationSensor?.let { manager.unregisterListener(this, it) }
      accelerometer?.let { manager.unregisterListener(this, it) }
      magnetometer?.let { manager.unregisterListener(this, it) }
    } else {
      manager.unregisterListener(this)
    }
    resetSensorState()
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
    eventSink = events
    start()
  }

  override fun onCancel(arguments: Any?) {
    stop()
    eventSink = null
  }
}